lang/src/org/partiql/lang/syntax/SqlParser.kt

```diff
@@ -2244,25 +2250,27 @@ class SqlParser(private val ion: IonSystem) : Parser {
         return ParseNode(ARG_LIST, null, items, rem)
     }
 
+
     /**
-     * Checks that the given [Token] list does not have any top-level tokens outside of the top level query. Throws
-     * an error if so.
-     *
-     * Currently only checks for `EXEC`. DDL and DML along with corresponding error codes to be added in the future
-     * (https://github.com/partiql/partiql-lang-kotlin/issues/354).
+     * Checks if the [ParseNode] requires to be only at top level. DML keyword tokens, for example, requires to be top level.
      */
-    private fun List<Token>.checkUnexpectedTopLevelToken() {
-        for ((index, token) in this.withIndex()) {
-            if (token.keywordText == "exec" && index != 0) {
-                token.err("EXEC call found at unexpected location", PARSE_EXEC_AT_UNEXPECTED_LOCATION)
-            }
+    private fun ParseNode.isTopLevelType() = type in listOf(UPDATE, DELETE, INSERT, INSERT_VALUE, REMOVE, EXEC)
+
+    /**
+     * Validates tree to make sure that the top level tokens are not found below the top level
+     */
+    private fun validateTopLevelNodes(node: ParseNode, level: Int) {
+        // Note that for DML operations, top level parse node is of type 'FROM'. Hence the check level > 1
```

Comment:
super nit: the top level parse node **could** be of type `FROM` (not necessary for all DML ops)

Reply:
It is for all DML ops if I am not wrong. I have updated the logic to take this into consideration.

Reply:
From the [partiql.ion](https://github.com/partiql/partiql-lang-kotlin/blob/master/lang/resources/org/partiql/type-domains/partiql.ion#L18-L26) grammar for DML ops, the `FROM` source is optional.